compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java

```diff
@@ -12005,29 +12016,29 @@ private STNode parseIntermediateClause(boolean isRhsExpr, boolean allowActions)
             case CONFLICT_KEYWORD:
                 return null;
             default:
+                if (nextToken.kind == SyntaxKind.IDENTIFIER_TOKEN && nextToken instanceof STIdentifierToken) {
+                    if (((STIdentifierToken) nextToken).text.equals(SyntaxKind.COLLECT_KEYWORD.stringValue())) {
+                        return parseCollectClause(isRhsExpr);
+                    }
+                    if (((STIdentifierToken) nextToken).text.equals(SyntaxKind.GROUP_KEYWORD.stringValue())) {
+                        return parseGroupByClause(isRhsExpr);
+                    }
+                }
                 recover(peek(), ParserRuleContext.QUERY_PIPELINE_RHS);
                 return parseIntermediateClause(isRhsExpr, allowActions);
         }
     }
 
     private STNode parseCollectClause(boolean isRhsExpr) {
         startContext(ParserRuleContext.COLLECT_CLAUSE);
-        STNode collectKeyword = parseCollectKeyword();
+        STToken nextToken = consume();
+        STNode collectKeyword = STNodeFactory.createToken(SyntaxKind.COLLECT_KEYWORD,
+                            nextToken.leadingMinutiae(), nextToken.trailingMinutiae());
```

Comment:
Let's have this inside the `parseCollectKeyword()`. We want to avoid exposing the underlying complexity as much as possible for readability/maintenance. 

We have a similar sample here for `KEY` keyword.
https://github.com/ballerina-platform/ballerina-lang/blob/442ecbe9f7fcf8aee80175c12dd118528fe6226f/compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java#L10680-L10698